File: backend/schemas/seat.py
"""Pydantic schemas for Seat model."""

from __future__ import annotations

from pydantic import AwareDatetime, Field

from backend.models._enums import SeatStatus
from backend.schemas.base import BaseCreateSchema, BaseResponseSchema


class SeatBase(BaseCreateSchema):
    name: str = Field(..., max_length=255)
    zone_id: str
    mac_address: str | None = Field(None, max_length=17)
    status: SeatStatus = SeatStatus.AVAILABLE
    plug_id: str | None = Field(None, max_length=255)
    is_console: bool = False
    notes: str | None = Field(None, max_length=1000)


class SeatCreate(SeatBase):
    pass


class SeatUpdate(BaseCreateSchema):
    name: str | None = Field(None, max_length=255)
    zone_id: str | None = None
    mac_address: str | None = Field(None, max_length=17)
    status: SeatStatus | None = None
    plug_id: str | None = Field(None, max_length=255)
    is_console: bool | None = None
    notes: str | None = Field(None, max_length=1000)


class SeatResponse(SeatBase, BaseResponseSchema):
    id: str
    created_at: AwareDatetime
    updated_at: AwareDatetime
    wol_attempts: int = 0
    wol_successes: int = 0
    wol_failures: int = 0

File: frontend/src/types/seat.ts
/** Seat status values from backend SeatStatus enum. */
export const SeatStatus = {
  AVAILABLE: 'AVAILABLE',
  ONLINE: 'ONLINE',
  IN_USE: 'IN_USE',
  RESERVED: 'RESERVED',
  PAUSED: 'PAUSED',
  MAINTENANCE: 'MAINTENANCE',
  OFFLINE: 'OFFLINE',
  BOOTING: 'BOOTING',
  UNREACHABLE: 'UNREACHABLE',
  EXPIRED: 'EXPIRED',
} as const;

export type SeatStatus = (typeof SeatStatus)[keyof typeof SeatStatus];

/** Full Seat entity as returned by `GET /api/seats`. */
export interface Seat {
  id: string;
  name: string;
  zone_id: string;
  zone_name?: string;
  mac_address: string | null;
  status: SeatStatus;
  plug_id: string | null;
  is_console: boolean;
  notes: string | null;
  overlay_forced: boolean;
  assigned_end_at: string | null;  // Epic 6.5.4: active session's assigned expiry; null when no limit set
  /** Maintenance downtime tracking (C.11): UTC ISO when the seat was marked
   *  MAINTENANCE; null when not under maintenance. */
  maintenance_since: string | null;
  /** Live downtime in seconds, computed server-side while MAINTENANCE. */
  maintenance_duration_seconds: number | null;
  wol_attempts: number;
  wol_successes: number;
  wol_failures: number;
  /** Active session ID, populated via WebSocket `seat_updated` events. */
  current_session_id?: string;
  /** Active session start time (ISO), used by the elapsed timer. Populated by
   *  GET /api/seats from the active session on the seat. */
  current_session_started_at?: string;
  created_at: string;
  updated_at: string;
}
